nntrainer/include/network_graph.h

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file    network_graph.h
+ * @date    19 Oct 2020
+ * @see     https://github.com/nnstreamer/nntrainer
+ * @author  Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug     No known bugs except for NYI items
+ * @brief   This is Network Graph Class for Neural Network
+ *
+ */
+
+#ifndef __NETWORK_GRAPH_H_
+#define __NETWORK_GRAPH_H_
+#ifdef __cplusplus
+
+#include <iostream>
+#include <layer_internal.h>
+#include <list>
+#include <loss_layer.h>
+#include <memory>
+#include <stack>
+#include <vector>
+
+namespace nntrainer {
+
+struct LayerNode {
+  std::shared_ptr<Layer> layer;
+  unsigned int index;
+};
+
+class NetworkGraph {
+public:
+  NetworkGraph() : num_node(0), def_name_count(0){};
+
+  void addEdge(unsigned int ith, LayerNode node);
+
+  void topologicalSort();
+
+  void addLayerNode(std::shared_ptr<Layer>);
+
+  unsigned int getNumNode() { return num_node; }
+
+  LayerNode &getLayerNode(unsigned int ith);
+
+  LayerNode &getLayerNode(const std::string layer_name);
+
+  void ensureName(std::shared_ptr<Layer> layer, const std::string &prefix,
+                  bool force_rename);
+
+  int setGraphNode(std::vector<std::shared_ptr<Layer>> layers,
+                   const LossType loss_type);
+
+  int realizeMultiInputType(Layer &current);
+
+  int realizeActivationType(Layer &current,
+                            std::vector<std::shared_ptr<Layer>> layers);
+
+  int realizeFlattenType(Layer &current,
+                         std::vector<std::shared_ptr<Layer>> layers);
+
+  int addLossLayer(const LossType loss_type);
+
+  int setEdge();
+
+private:
+  void topologicalSortUtil(unsigned int ith, bool visited[],
+                           std::stack<LayerNode> &Stack);
+
+  unsigned int num_node;
+  std::vector<std::list<LayerNode>> adj;
+  std::vector<LayerNode> Sorted;
+  std::set<std::string> layer_names;
```

Comment:
so these things will move out of the neuralnetwork class right?

Reply:
These are not exposed by neural network I guess. after compile, neural network just call forwarding or backwarding of graph I guess.

Reply:
Graph must expose functions which provides an iterator view of the graph. NeuralNetwork will choose which functions to call or directly make the call.

Reply:
> Graph must expose functions which provides an iterator view of the graph. NeuralNetwork will choose which functions to call or directly make the call.

I think some of functions only to display the graph structure. There are many layers included by network graph. Like out layer, addition layers etc. So, eventually yes, we have to expose graph. But not all I say. And Also I'm not sure what you mean choose which functions?? I mean what do you expect which functions to choose?

Reply:
By choose I meant if forwarding to be run or backwarding to be run will be chosen by NeuralNetwork. Depending on train/inference, NeuralNetwork class can choose/decide.